File: c/stl/stack.c
// STL stack
#include <iostream>
#include <stack>


using namespace std;

int main()

{
  stack<char> st;

  st.push('A');
  st.push('B');
  st.push('C');
  st.push('D');

  for(; !st.empty(); st.pop()){

    cout << "\nPopping: ";
    cout << st.top();

  }

  cout << endl;



  return 0;

}

File: haskell/Stack.hs
module Stack (Stack, push, peek, pop, isEmpty, empty, toList) where

newtype Stack a = MkStack [a]

empty :: Stack a
empty = MkStack []

push :: Stack a -> a -> Stack a
push (MkStack xs)  x = MkStack (x:xs)

peek :: Stack a -> Maybe a
peek (MkStack [])  = Nothing
peek (MkStack (x:xs)) = Just x

isEmpty :: Stack a -> Bool
isEmpty (MkStack []) = True
isEmpty _  = False

pop :: Stack a -> Stack a
pop (MkStack []) = empty
pop (MkStack (x:xs)) = MkStack xs

toList :: Stack a -> [a]
toList (MkStack xs) = xs

instance Show a => Show (Stack a) where
  show q = show (toList q)



